crochet-stitcher/src/stages/1-parser.ts

```diff
@@ -1,5 +1,277 @@
+                val = val.substring(1);
+                continue;
+            }
+            if (val[0] == '{') {
+                output.push({ type: TokenType.LParen, value: '{' });
+                val = val.substring(1);
+                continue;
+            }
+            if (val[0] == '[') {
+                output.push({ type: TokenType.LParen, value: '{' });
+                val = val.substring(1);
+                continue;
+            }
+            //strip off trailing colons, rparens, or commas (only type of trailing according to spec)
+            if (val.slice(-1) == ')') {
+                endTokens.push({ type: TokenType.RParen, value: ')' });
+                val = val.substring(0, val.length - 1);
+                continue;
+            }
+            if (val.slice(-1) == '}') {
+                endTokens.push({ type: TokenType.RParen, value: '}' });
+                val = val.substring(0, val.length - 1);
+                continue;
+            }
+            if (val.slice(-1) == ']') {
+                endTokens.push({ type: TokenType.RParen, value: '}' });
+                val = val.substring(0, val.length - 1);
+                continue;
+            }
+            if (val.slice(-1) == ':') {
+                endTokens.push({ type: TokenType.Colon, value: ':' });
+                val = val.substring(0, val.length - 1);
+                continue;
+            }
+            if (val.slice(-1) == ',') {
+                endTokens.push({ type: TokenType.Comma, value: ',' });
+                val = val.substring(0, val.length - 1);
+                continue;
+            }
+            // console.log(`sliced ${tokens[i]} to ${val}`)
+            break;
+        }
+
+        // TODO: Split strings here (ch20 is valid technically)
+
+        // So let's try parsing it as an int
+        let numberCandidate = parseInt(val);
+        if (Number.isNaN(numberCandidate)) {
+            // LUT for valid strings
+            endTokens.push(stringTokenLUT(val));
+        } else {
+            // If it is a negative number throw exn
+            if (numberCandidate < 1) {
+                throw new RangeError(`Repeat counts or stitch counts must be positive integers! 
+                                        Found zero or negative value ${numberCandidate}`);
+            }
+            endTokens.push({ type: TokenType.Number, value: numberCandidate });
+        }
+
+        for (let i = endTokens.length - 1; i >= 0; i--) {
+            output.push(endTokens[i]);
+        }
+    }
+    return output;
+}
 
 export function parse(input: string): Pattern<ParsedStitch> {
-    throw 0; // TODO
+    let tokens = tokenizer(input);
+    console.log(tokens);
+    // console.log("Completed tokenization")
+    let foundation = Foundation.SlipKnot;
+    let colour = Colour.White;
+
+    let stitches: ParsedStitch[] = [];
+    let curStitchType = null;
+    let curStitchCount = null;
```

Comment:
You should add types to these, since TypeScript can't infer the type from just `null`.

```js
let curStitchType: StitchType | null = null;
```